### ISSUE TITLE
Some questions about Figure 4？

### ISSUE THREAD

**Issue by adreamero** [takashiishida/flooding#5]
How did you get the amplitude of test loss gradient under each epoch in Figure 4 ? Isn't there no backpropagation when calculating test loss under each epoch?

**Comment by takashiishida** [comment]
We had to do backpropagation additionally so that we can get the amplitude of test loss gradient.

An example:

```python
def get_grads(loader, model, device):
    num_samples = 0
    criterion = nn.CrossEntropyLoss(reduction='sum')
    model.eval()

    sum_grads = []
    model.zero_grad()
    for param in model.parameters():
        if not param.requires_grad:
            continue
        sum_grads.append(None)

    for (images, labels, *empty) in loader:
        images, labels =images.to(device), labels.to(device)
        outputs = model(images)
        loss = criterion(outputs, labels.long())
        model.zero_grad()
        loss.backward()
        it_params = (p for p in model.parameters() if p.requires_grad)
        for i, param in enumerate(it_params):
            if sum_grads[i] is None:
                sum_grads[i] = param.grad
            else:
                sum_grads[i] += param.grad
        num_samples += labels.size(0)

    avg_grads = []
    for sum_grad in sum_grads:
        avg_grads.append(sum_grad.view(-1) / num_samples)

    return avg_grads
```

**Comment by adreamero** [comment]
Thank you very much for your reply, which is basically consistent with what I think, haha
